README.md
# validator [![spm version](https://moekit.com/badge/validator)](https://moekit.com/package/validator)

---

## 说明
> 该模块只用来做表单验证，提供内置验证功能，支持自定义验证事件

## 使用方法
> 1. 存在`data-valid`or`name`属性则有验证规则
> 2. 存在`required`属性，则必须验证规则
> 3. 存在`pattern`属性，同时存在`data-valid`属性时，每次都会用`pattern`的正则规则来验证
> 4. 存在`min`、`max`属性，用于验证长度规则（中文算两个字符）
> 5. 存在`data-alt`属性，验证失败的文案，如果当前标签的sibling存在`validator-info`类名时，该文案会赋值到此sibling中
> 6. 存在`data-hint`属性，验证成功的文案，如果当前标签的sibling存在`validator-info`类名时，该文案会赋值到此sibling中
> 7. 存在`type`属性，且其值为`form`时，非表单元素可以模拟表单元素进行校验（如: div）

```html
<div id="JS_form">
	<input class="validator" date-valid="qq" pattern="^\d*$"
		 min="5" max="11" name="qq" data-alt="qq必须是5-11位数字" type="text" />
	<div class="simulate-input" name="myname" value="" type="form" required>我是一个虚拟的表单元素</div>
	<span class="validator-info"></span>
	<button id="JS_submit">提交</button>
</div>
```

## 初始化参数
+ id: 表单id，取不到id对应的dom时默认使用document
+ custom: `Object`用于自定义表单事件，其key值对应验证标签中的[`data-valid`属性]
+ showOnClass：文案显示在类名为showOnClass的sibling上，默认validator-info
+ isFocus: 验证失败后，焦点是否跳到失败节点上，默认true
+ pass: 调试使用，true时直接跳过验证，直接进入验证通过事件，默认false

## 事件机制
+ `on`事件，用于绑定`验证失败事件`，`验证成功事件`
	+ `error` 验证失败回调事件，带参数`function`，其中function返回值有name, dom, alt
	+ `success` 验证通过回调事件，带参数`function`
+ `onvalid`事件，用于绑定`自定义验证事件`，带3个参数
	+ Param 1: `string`自定义验证事件的名称，可用于验证标签中的[`data-valid`属性]
	+ Param 2: `function`函数，其中function带有形参val,dom; 用于绑定自定义验证事件函数，此函数必须带有返回值为true or false
	+ Param 3: `function`函数，成功绑定事件的回调
+ `onvalid`事件，用于解绑`验证事件`，带2个参数
	+ Param 1: `string`验证事件的名称
	+ Param 2: `function`函数，成功绑定事件的回调
+ `verify`事件，用于开始执行验证，验证通过则调用`on`的`success`事件，反之调用`on`的`error`事件
	+ [Param 1]: `function`用于验证通过回调，会阻止`on`的`success`事件的调用
	+ [Param 2]: 调试使用，`true`时直接跳过验证，直接进入验证通过事件
+ `is`事件，简单对一个值进行验证
	+ Param 1: 验证类型，即data-valid的值，也支持自定义验证规则
	+ Param 2: 验证的值
	+ [Param 3]: 验证的值的字符长度最小值min
	+ [Param 4]: 验证的值的字符长度最大值max
	+ [Param 5]: 传入dom对象，专用于自定义验证规则可能使用到的对象

## data-valid属性值介绍
+ number: 只能是数字
+ int: 只能是整数
+ mobile: 手机号码格式
+ phone: 固定电话格式
+ mobilephone: 手机或固话格式
+ zipcode: 邮政编码格式
+ email: 邮箱格式
+ date: 日期格式`2015-01-01` or `2015/01/01`
+ time: 时间格式`11:56` or `11:56:29`
+ datetime: 日期时间格式`2015-01-01 11:56` or `2015-01-01 11:56:29`
+ idcard: 身份证格式 
+ url: 网址格式 
+ wechatid: 微信号（微信账号仅支持6-20个字母、数字、下划线或减号，以字母开头） 

## 规则优先级说明
绑定规则emit("onvalid", func) > pattern属性正则规则 > 初始化参数custom规则 > 内置正则规则

## 注意
+ select校验时，option的value必须设置，就算为空也要设置为空
+ 非表单元素校验时必须增加`type="form"`, 别忘了增加`required`和`name`属性才能校验

## @todo
+ select校验、非元素表单校验少了单元测试

HISTORY.md
# History

---

## 0.0.1
`new` It is the first version of seedit-validform.

## 0.0.2
`FIXED` 邮编验证正则修正

## 0.0.3

`FIXED` 处理pass和isFocus参数的truefalse判断

`NEW` 新增内置规则-微信号（wechatid）

## 0.0.4

`NEW` 增加表单元素select标签的校验

`NEW` 增加非表单元素的校验


tests/validator-spec.js
var expect = require('expect.js');
var $ = require("jquery");
var Validator = require('validator');
describe('validator', function() {

	it('参数为空', function() {
		var validator = new Validator();
	});

	it('参数是json格式且为空', function() {
		var validator = new Validator({});
	});
	it('参数是json, id = JS_validform', function() {
		var validator = new Validator({
			"id": "JS_validform"
		});
	});
	it('设置自定义验证事件', function() {
		var validator = new Validator({
			"custom": {
				"qq": "^\d{4,11}$"
			}
		});
	});
	it('设置校验失败时不自动获取焦点', function() {
		var validator = new Validator({
			"isFocus": false
		});
	});
	it('设置校验结果显示在类名中', function() {
		var validator = new Validator({
			"showOnClass": "validator-success-word"
		});
	});
	it('设置跳过校验过程，直接校验通过', function() {
		var validator = new Validator({
			"pass": true
		});
	});
	it('设置完整的参数,json格式', function() {
		var validator = new Validator({
			"id": "JS_validform",
			"custom": {
				"qq": "^\d{4,11}$"
			},
			"isFocus": false,
			"showOnClass": "validator-success-word",
			"pass": true
		});
	});
	it('开始校验表单', function() {
		new Validator().verify();
	});
	it('is事件直接校验', function() {
		expect(new Validator().is("number",1234)).to.be(true);
	});
	it('is事件直接校验，容错测试-参数为空', function() {
		expect(new Validator().is()).to.be(false);
	});
	it('is事件直接校验，容错测试-设置不存在校验类型', function() {
		expect(new Validator().is("fe")).to.be(false);
	});
	it('校验内置规则-number', function() {
		expect(new Validator().is("number", 20150702)).to.be(true);
		expect(new Validator().is("number", 20150702.2)).to.be(true);
		expect(new Validator().is("number", -20150702.2)).to.be(true);
		expect(new Validator().is("number", "string")).to.be(false);
		expect(new Validator().is("number", "string666")).to.be(false);
	});
	it('校验内置规则-int', function() {
		expect(new Validator().is("int", 20150702)).to.be(true);
		expect(new Validator().is("int", 20150702.2)).to.be(false);
	});
	it('校验内置规则-mobile', function() {
		expect(new Validator().is("mobile", 18620065910)).to.be(true);
		expect(new Validator().is("mobile", 12620065910)).to.be(false);
	});
	it('校验内置规则-phone', function() {
		expect(new Validator().is("phone", "020-89899291")).to.be(true);
		expect(new Validator().is("phone", "20-89899291")).to.be(false);
	});
	it('校验内置规则-mobilephone', function() {
		expect(new Validator().is("mobilephone", 18620065910)).to.be(true);
		expect(new Validator().is("mobilephone", "020-89899291")).to.be(true);
		expect(new Validator().is("mobilephone", 12220065910)).to.be(false);
		expect(new Validator().is("mobilephone", "20-8989929")).to.be(false);
	});
	it('校验内置规则-zipcode', function() {
		expect(new Validator().is("zipcode", 123456)).to.be(true);
		expect(new Validator().is("zipcode", 1234)).to.be(false);
	});
	it('校验内置规则-email', function() {
		expect(new Validator().is("email", "test1001@qq.com")).to.be(true);
		expect(new Validator().is("email", "test1001@qq")).to.be(false);
	});
	it('校验内置规则-date', function() {
		expect(new Validator().is("date", "2015-07-02")).to.be(true);
		expect(new Validator().is("date", "2015/07/02")).to.be(true);
		expect(new Validator().is("date", "2015+07-02")).to.be(false);
		expect(new Validator().is("date", "2015/07+02")).to.be(false);
	});
	it('校验内置规则-time', function() {
		expect(new Validator().is("time", "09:05:05")).to.be(true);
		expect(new Validator().is("time", "09:05-05")).to.be(false);
	});
	it('校验内置规则-datetime', function() {
		expect(new Validator().is("datetime", "2015-07-02 09:05:05")).to.be(true);
		expect(new Validator().is("datetime", "2015/07/02 09:05:05")).to.be(true);
		expect(new Validator().is("datetime", "2015+07+02 09:05:05")).to.be(false);
		expect(new Validator().is("datetime", "2015/07/02 09-05-05")).to.be(false);
	});
	it('校验内置规则-idcard', function() {
		expect(new Validator().is("idcard", 310000197308225797)).to.be(true);
		expect(new Validator().is("idcard", 414699080)).to.be(false);
	});
	it('校验内置规则-url', function() {
		expect(new Validator().is("url", "http://bbs.bozhong.com")).to.be(true);
		expect(new Validator().is("url", "iamurl")).to.be(false);
	});
	it('校验内置规则-wechatid', function() {
		expect(new Validator().is("wechatid", "a123456")).to.be(true);
		expect(new Validator().is("wechatid", "A1234w6")).to.be(true);
		expect(new Validator().is("wechatid", "a123_456")).to.be(true);
		expect(new Validator().is("wechatid", "a123-456")).to.be(true);
		expect(new Validator().is("wechatid", "a1_3-456")).to.be(true);
		expect(new Validator().is("wechatid", "123a567")).to.be(false);
		expect(new Validator().is("wechatid", "12-4567")).to.be(false);
		expect(new Validator().is("wechatid", "12345_7")).to.be(false);
	});
	it('设置并校验min,max规则', function() {
		$("body").append('<div id="JS_min_max" style="display: none;"><input type="text" name="account" min="3" max="7" value="12345" required /></div>');
		new Validator({
			"id": "JS_min_max"
		}).on("error", function(name) {
			expect(false).to.be(true);
		}).verify(function(){
			expect(true).to.be(true);
		});
		$("body").append('<div id="JS_min_max_1" style="display: none;"><input type="text" name="account" min="3" max="7" value="123456789" required /></div>');
		new Validator({
			"id": "JS_min_max_1"
		}).on("error", function(name) {
			expect(true).to.be(true);
		}).verify(function(){
			expect(false).to.be(true);
		});
	});
	it('设置并校验pattern规则', function() {
		$("body").append('<div id="JS_pattern" style="display: none;"><input type="text" name="account" data-valid="account" pattern="^[a-zA-Z0-9]{3,7}$" value="12345" required /></div>');
		new Validator({
			"id": "JS_pattern"
		}).on("error", function(name) {
			expect(false).to.be(true);
		}).verify(function(){
			expect(true).to.be(true);
		});
		$("body").append('<div id="JS_pattern_1" style="display: none;"><input type="text" name="account" data-valid="account" pattern="^[a-zA-Z0-9]{3,7}$" value="123456789" required /></div>');
		new Validator({
			"id": "JS_pattern_1"
		}).on("error", function(name) {
			expect(true).to.be(true);
		}).verify(function(){
			expect(false).to.be(true);
		});
	});

	it('绑定自定义校验事件', function() {
		new Validator().onvalid("qq", function(val, dom) {
			return /^\d{4,11}$/.test(val);
		});
	});
	it('校验自定义校验事件', function() {
		var validator = new Validator().onvalid("qq", function(val, dom) {
			return /^\d{4,11}$/.test(val);
		});
		expect(validator.is("qq", 414699080)).to.be(true);
		expect(validator.is("qq", 310000197308225797)).to.be(false);
		var validator_1 = new Validator({
			"custom": {
				"qq": function(val) {
					return /^\d{4,11}$/.test(val);
				}
			}
		});
		expect(validator_1.is("qq", "3344556677")).to.be(true);
		expect(validator_1.is("qq", "33445566778899001122")).to.be(false);
	});
	it('解绑校验事件并校验', function() {
		expect(new Validator().is("number", 1234)).to.be(true);
		expect(new Validator().offvalid("number").is("number", 1234)).to.be(false);
	});
	it('设置校验通过回调', function() {
		new Validator().on("success", function() {});
	});
	it('设置校验通过回调', function() {
		new Validator().on("error", function() {});
	});
	it('触发校验通过的回调事件', function() {
		$("body").append('<div id="JS_trigger_success" style="display: none;"></div>');
		new Validator({
			"id": "JS_trigger_success"
		}).on("success", function() {
			expect(true).to.be(true);
		}).on("error", function() {
			expect(false).to.be(true);
		}).verify();
	});
	it('触发校验失败的回调事件', function() {
		$("body").append('<div id="JS_trigger_error" style="display: none;"><input name="test" required /></div>');
		new Validator({
			"id": "JS_trigger_error"
		}).on("success", function() {
			expect(false).to.be(true);
		}).on("error", function() {
			expect(true).to.be(true);
		}).verify();
	});
	it('使用链式事件', function() {
		new Validator()
			.onvalid("qq", function() {
				return true;
			})
			.offvalid("int")
			.on("success", function() {})
			.on("error", function() {})
			.verify()
			.is("number", 1234);
	});
});

index.js
"use strict";
var Events = require('eventor');
var Validator = function(option){
	return this.init(option)
}
Events.mixTo(Validator)
Validator.prototype.init = function(option){
	var _this = this
	option = option || {}
	this.id = option.id || ''
	this.dom = document.getElementById(this.id) || document
	this.validator = this.dom.querySelectorAll('[data-valid], [name]')
	this.custom = {}
	this.showOnClass = option.showOnClass || "validator-info"
	this.pass = option.pass === true ? true : false
	this.isFocus = option.isFocus === false ? false : true
	this.reg = {
		number: /^-?\d+(\.\d+)?$/, // 数字 22, 22.2 => true; 22. => false;
		int: /^-?\d+$/, // 整数
		mobile: /^1[3,4,5,7,8][0-9]{9}$/, // 手机号码 13800138000
		phone: /^((\d{3}|\d{4})-(\d{7}|\d{8}))|((\d{3}|\d{4})-(\d{7}|\d{8})-\d{4})$/, // 电话号码 020-34343434
		mobilephone: /^(1[3,4,5,7,8][0-9]{9})|((\d{3}|\d{4})-(\d{7}|\d{8}))|((\d{3}|\d{4})-(\d{7}|\d{8})-\d{4})$/,
		zipcode: /^\d{6}$/, // 邮政编码 560310, 056300
		email: /^(\w)+(\.\w+)*@(\w)+((\.\w+)+)$/, // 邮箱 123456@qq.com
		date: /^(19\d{2}|2\d{3})[-\/](0[0-9]|1[0-2])[-\/]([0-2][0-9]|3[0-1])$/, // 日期-年月日 2015-02-03 2015/02/03
		time: /^(([0-1][0-9])|(2[0-3]))(:[0-5][0-9]){1,2}$/, // 日期-时分秒 12:50 12:50:23
		datetime: /^(19\d{2}|2\d{3})[-\/](0[0-9]|1[0-2])[-\/]([0-2][0-9]|3[0-1])\s([0-1][0-9]|2[0-3])(:[0-5][0-9]){1,2}$/,
		idcard: /^\d{18}|(\d{17}(\d|X|x))$/, // 只允许18位的身份证
		url: /^((http|ftp|https):\/\/)?[\w\-_]+(\.[\w\-_]+)+([\w\-\.,@?^=%&amp;:\/~\+#]*[\w\-\@?^=%&amp;\/~\+#]?)$/,
		wechatid: /^[a-zA-Z][a-zA-Z0-9_\-]{5,19}$/
	}
	// reg转为custom 
	for( var i in _this.reg ){
		_this.custom[i] = (function(){
			var key = i;
			return function(val, dom){ return _this.reg[key].test(val) }
		}())
	}
	// 初始化参数custom存入custom
	if( option.custom instanceof Object ){
		for( var i in option.custom ){
			this.custom[i] = option.custom[i];
		}
	}
	// pattern存入type并转为custom
	var pattern = this.dom.querySelectorAll('[pattern]');
	for(var i=0; i<pattern.length; i++){
		var type = pattern[i].getAttribute("data-valid");
		if( !!type ){
			this.reg[type] = new RegExp(pattern[i].getAttribute("pattern"));
			this.custom[type] = (function(){
				var key = type;
				return function(val, dom){ return _this.reg[key].test(val) }
			}())
		}
	}
	// 收集自定义事件
	// this.on("onvalid", function(_Type, cb){
	// 	_this.custom[_Type] = cb
	// });
	// this.on("offvalid", function(_Type){
	// 	_this.custom[_Type] && delete _this.custom[_Type]
	// });

	// for( var i=0; i<_this.validator.length; i++ ){
	// 	_this.validator[i].addEventListener("blur",function(){
	// 		_this.verifyOne(this,false);
	// 	},false);
	// }
	this.eventListener();
	return this
}
// 如果存在兄弟类名showOnClass，则获取焦点时移除showOnClass的值
Validator.prototype.eventListener = function(){
	var _this = this;
	for( var i=0; i<=_this.validator.length-1; i++ ){
		if( (!!_this.validator[i].getAttribute("data-alt") || !!_this.validator[i].getAttribute("data-hint")) 
				&& !!_this.validator[i].parentNode.querySelector("."+_this.showOnClass) ){
			_this.validator[i].ind = i;
			_this.validator[i].addEventListener("focus", function(){
				_this.validator[this.ind].parentNode.querySelector("."+_this.showOnClass).innerText = "";
			}, false);
		}
	}
	return _this;
}
// 收集自定义事件
Validator.prototype.onvalid = function(_Type, cb, cb1){
	this.custom[_Type] = cb;
	this.custom[_Type] && cb1 && cb1();
	return this;
}
// 移除自定义事件
Validator.prototype.offvalid = function(_Type, cb){
	this.custom[_Type] && delete this.custom[_Type] && cb && cb();
	return this;
}
Validator.prototype.verify = function(cb, bool){
	try{
		var _this = this;
		for( var i=0; i<=_this.validator.length; i++ ){
			// 验证直至最后一个没有异常退出，证明验证通过，执行success
			if( i == _this.validator.length || bool === true || _this.pass === true ){
				!!cb ? cb() : _this.emit("success")
				break
			}
			if( _this.verifyOne( _this.validator[i], true ) ){
				break
			}
		}
		return _this;
	} catch(err){
		throw err;
	}
}
// 对每一个表单元素进行验证
Validator.prototype.verifyOne = function(verify_dom, isFocus){
	var _this = this;
	// 是否需要验证
	var _Required = verify_dom.getAttribute("required")
	var _Required = _Required === "" ? "required" : _Required;
	// 验证类型
	var _Type = verify_dom.getAttribute("data-valid")
	// 验证便签字段
	var _Field = verify_dom.getAttribute("name") || "name"
	// 验证的值
	var _Val = verify_dom.getAttribute("type") === "form" ? verify_dom.getAttribute("value") : verify_dom.value;
	// 验证的值的最小长度
	var _Min = verify_dom.getAttribute("min")
	// 验证的值的最大长度
	var _Max = verify_dom.getAttribute("max")
	// 验证失败显示的文案
	var _Alt = verify_dom.getAttribute("data-alt")
	// 验证成功显示的文案
	var _Hint = verify_dom.getAttribute("data-hint")
	// 自定义正则验证
	var _Pattern = verify_dom.getAttribute("pattern")

	/*
	 * 表单验证不通过的条件，如下：
	 */
	var dom = verify_dom.parentNode.querySelectorAll("."+_this.showOnClass);
	if( (!!_Required || !!_Val) && !_this.requiredVerify(_Type, _Val, _Min, _Max, verify_dom) ){
		
		if( !!_Alt ){
			if( dom.length > 0 && !!dom[0] ){
				dom[0].raw = (dom[0].raw === "" || (!!dom[0].raw && dom[0].raw.length>0)) ? dom[0].raw : dom[0].innerText;
				dom[0].innerText = _Alt;
			} else {
				_this.emit("error", _Field, verify_dom, _Alt)
				if( _this.isFocus === true && isFocus === true ){ verify_dom.focus() }
			}
		} else {
			_this.emit("error", _Field, verify_dom) 
			if( _this.isFocus === true && isFocus === true ){ verify_dom.focus() }
		}
		return true
	} else {
		if( dom.length > 0 && !!dom[0] ){
			dom[0].raw = !!dom[0].raw ? dom[0].raw : dom[0].innerText;
			dom[0].innerText = _Hint || (dom[0].raw || "");
		}
	}
	return false
}
Validator.prototype.requiredVerify = function(type, val, min, max, dom){
	if( !!type && !!this.custom[type] ){
		if( !!this.custom[type](val, dom) ){
			return isBelongRange(val,min,max) ? true : false;
		} else return false;
	} else {
		if( !!val ){
			return isBelongRange(val,min,max) ? true : false;
		} else return false;
	}
	function isBelongRange(val, min, max){
		if( !min && !max ) return true;
		val = val.replace(/[^\x00-\xff]/g,"**");
		return (!!min && !!max ? (val.length>=min && val.length<=max?true: false) : ((!!min&&val.length>=min) || (!!max&&val.length<=max)?true: false) )
	}
}
Validator.prototype.is = function(type, val, min, max, dom){
	if( !this.custom[type] ){
		if( window.console) console.log(type+"对应的验证规则不存在");
		return false;
	}
	return this.requiredVerify(type, val, min, max, dom);
}
module.exports = Validator